compiler/lib/deadcode.ml

```diff
@@ -109,8 +123,8 @@ and mark_reachable st pc =
 let live_instr st i =
   match i with
   | Let (x, e) -> st.live.(Var.idx x) > 0 || not (pure_expr st.pure_funs e)
-  | Assign (x, _) -> st.live.(Var.idx x) > 0
-  | Set_field _ | Offset_ref _ | Array_set _ -> true
+  | Assign (x, _) | Set_field (x, _, _) -> st.live.(Var.idx x) > 0
+  | Offset_ref _ | Array_set _ -> true
```

Comment:
Why don't we have the same logic for `Array_set` ? (when we statically know that bounds are ok)

Reply:
Using a forward reference is a common pattern which is worthwhile to take into account. I'm not sure there is much to gain with `Array_set` and this adds some complexity.